File: Data Structures & Algorithms/single-number/submission-0.cpp
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        int result = nums[0];
        for (int i = 1; i < nums.size(); i++) {
            result = result ^ nums[i];
        }

        return result;
    }
};

// 7, 6, 6, 7, 8
// 0010 XOR 0010 = 0000
// all XOR will cancel out

File: Data Structures & Algorithms/binary-search/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1

        while l <= r:
            m = (l + r) // 2

            if nums[m] == target:
                return m
            elif nums[m] > target:
                r = m - 1
            else:
                l = m + 1

        return -1
        

# [-1, 0, 2, 4, 6, 8], target = 4
#   l              r
#         m

# [-1, 0, 2, 4, 6, 8], target = 4
#            l     r
#               m

# [-1, 0, 2, 4, 6, 8], target = 4
#            l     
#            r  
#            m 